src/openfermion/measurements/_qubit_partitioning.py

```diff
@@ -0,0 +1,141 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+
+""" Code to generate Pauli strings for measurement of local operators"""
+import numpy
+from itertools import chain
+try:
+    from itertools import zip_longest
+except ImportError:
+    from itertools import izip_longest as zip_longest
+
+
+def binary_partition_iterator(qubit_list, num_partitions=None):
+    '''Generator for a list of 2-partitions of N qubits
+    such that all pairs of qubits are split in at least one partition,
+    following ArXiv:1908.05628
```

Comment:
The code in this function does not follow the paper as far as I can tell; the scheme in the paper is based on using the binary representations of the qubit indices. Is there a reason you don't follow the paper? It's difficult for me to verify the correctness of the algorithm you give here.

Reply:
Sorry for not detailing the difference - I coded it this way because the output is more balanced, and because this zipping method seems faster/more pythonic than trying to deal with binary division in python.

The code does follow the paper in the case where len(qubit_list) = 2^n. By cutting the list in two and zipping it back together we in effect shuffle the indices.

For example, consider the list [1, 2, 3, 4, 5, 6, 7, 8]. Both the code and the paper first iterate to generate [[1, 2, 3, 4], [5, 6, 7, 8]].
Then, in the paper, we should take the second index, i.e. [[1, 2] + [5, 6], [3, 4] + [7, 8]].
In the code, we zip the list together to generate [1, 5, 2, 6, 3, 7, 4, 8], and then split this again in 2 to obtain [[1, 5, 2, 6], [3, 7, 4, 8]]. As we don't care about order in the lists, the two sets are equivalent.

When len(qubit_list) != 2^n, the code and the paper differ, but the result of the code is better as it is more balanced. For example, consider the list [1, 2, 3, 4, 5] - the paper would want to split this initially into [[1, 2, 3, 4], [5]], while the code would split this into [[1, 2, 3], [4, 5]]. The former would be less preferable in an experiment as we wind up making more repeat measurements of the form XX/YY/ZZ on pairs of qubits.

The new code also has the nice effect for k>2 that the partitions are ordered 'correctly' for the sub-partitioning in Fig.1(b) of the paper. For example, if I look again at [1,2,3,4,5,6,7,8], after partitioning into [[1, 2] + [5, 6], [3, 4] + [7, 8]], the scheme in the paper requires me to partition [1, 2, 5, 6] into [1] + [5], [2] + [6]. In the code, as our list is ordered already as [1, 5, 2, 6], we can repeat the direct division into half.

I've updated the tests to check that the code behaves appropriately for k=3 and N=1 to 15 (previously it only checked for k=3, N=8) - this is the 'test_partition_three' function in the test suite. This checks that for each index [i, j, k] we find one partition that exactly divides them, as required. I've also added some comments for clarity. Hope this helps.